picard/util/__init__.py

```diff
@@ -88,6 +88,19 @@ def unlock(self):
         self.__lock.unlock()
 
 
+def process_events_iter(iterable, steps=10):
+    """
+    Creates an iterator over iterable that calls QCoreApplication.processEvents()
+    in certain iteration steps.
+
+    This must only be used in the main thread.
+    """
+    for i, item in enumerate(iterable):
+        if (i + 1) % steps == 0:
+            QtCore.QCoreApplication.processEvents()
+        yield item
```

Comment:
Using `steps` has low overhead, but it's hard to estimate the correct value, as each step duration is not really predictable (as it depends on many things, like code executed and hardware the code runs on).

I wonder about acceptable overhead and if we could use a timer.
Something like:

```python
from time import monotonic, sleep


def process_events_iter(iterable, interval=0.1):
    if interval:
        start = monotonic()
    for item in iterable:
        if interval:
            now = monotonic()
            delta = now - start
            if delta > interval:
                start = now
                print("call QtCore.QCoreApplication.processEvents()")
        yield item


for i in process_events_iter(range(0, 20), interval=0.15):
    print(i)
    sleep(0.05)

for i in process_events_iter(range(0, 20), interval=0.4):
    print(i)
    sleep(0.05)
```

Since code processed in each step is likely to be much slower than call to `monotonic()` it could be acceptabe. What do you think?

Reply:
Let's try it

Reply:
I don't get satisfying results with the timer approach. When loading the files the UI stotters a lot initially. I could not find any interval that works well. I think we should stick with the counter.

Reply:
> I don't get satisfying results with the timer approach. When loading the files the UI stotters a lot initially. I could not find any interval that works well. I think we should stick with the counter.

That's weird.
I did some tests, and it seems to me the reason is because:
- when iteration is very short, processEvents() isn't called (because delta > interval doesn't happen)

Here is a modified test code:

```python
def process_events_iter(iterable, steps=None, interval=0.1):
    """
    Creates an iterator over iterable that calls QCoreApplication.processEvents()
    in certain iteration steps.

    This must only be used in the main thread.
    """
    if interval:
        start = monotonic()
    called = False
    _id = id(iterable)
    print("start id=%d len=%d interval=%f" % (_id, len(iterable), interval))

    def process_events():
        print("===== call QtCore.QCoreApplication.processEvents() ====")
        QtCore.QCoreApplication.processEvents()

    for item in iterable:
        if interval:
            now = monotonic()
            delta = now - start
            print("id=%d delta=%f" % (_id, delta))
            if delta > interval:
                start = now
                process_events()
                called = True
        yield item
    if not called:
        process_events()
```

It calls `process_events()` at least once at end of iteration, or every 0.1 seconds.
I tested with 1370 files moved to clusters.

Reply:
I would not bother keeping track of whether processEvents() was ever called, but instead to simply call it regardless at the end of the loop.